Update Air Quality tutorial (#4340)
viamrobotics/docs: docs/tutorials/control/air-quality-fleet.md

# 4. The reader can deploy custom frond ends that end users can use to operate their machines.
---

In this tutorial you will learn how to set up a fleet of devices for yourself or third parties to collect air quality data.
You will then create a web app that shows the most recent reading for any device a user has access to.

{{< alert title="Learning Goals" color="info" >}}

By completing this project, you will learn to:

- Configure a fleet of machines
- Organize your fleet using {{< glossary_tooltip term_id="location" text="locations" >}}
- Collect and sync data from multiple machines
- Use the Viam TypeScript SDK to query sensor data
- Create a custom dashboard that you and third parties can use to view data for their respective machines.

{{< /alert >}}

{{<imgproc src="/tutorials/air-quality-fleet/three-sensor-dash-wide.png" resize="1000x" style="width:500px" declaredimensions=true alt="Air quality dashboard in a web browser with PM2.5 readings from three different sensor machines displayed." class="imgzoom">}}

## Requirements

You can create one or more machines to measure air quality.
For each machine, you need the following hardware:

> - one or more [SDS011 Nova PM sensors](https://www.amazon.com/SDS011-Quality-Detection-Conditioning-Monitor/dp/B07FSDMRR5)

Author: ```suggestion
- One [SDS011 Nova PM sensors](https://www.amazon.com/SDS011-Quality-Detection-Conditioning-Monitor/dp/B07FSDMRR5)
```

You have configured the sensor so the board can communicate with it, but sensor data is not yet being saved anywhere.
Viam's [data management service](/data-ai/capture-data/capture-sync/) lets you capture data locally from each sensor and then sync it to the cloud where you can access historical sensor data and see trends over time.
As you configure more sensing machines, you'll be able to remotely access data from all machines.

{{< table >}}

{{% tablestep number=1 %}}

Click **+** and add the **data management** service.

On the data manager panel:

- Toggle **Syncing** to the on position.
- Set the sync interval to `0.05` minutes (every 3 seconds).
- In the **Tags** field, add `air-quality`.
  This tag will now automatically be applied to all data collected by this data manager which will make querying data easier.

{{% /tablestep %}}
{{% tablestep number=2 %}}

On the **PM_sensor** panel, click **Add method** to add data capture.

- **Type**: **Readings**.
- **Frequency**: `0.1` (every 10 seconds).

Save the config.

You can check that your sensor data is being synced by clicking on the **...** menu and clicking **View captured data**.

{{% /tablestep %}}
{{< /table >}}

> Congratulations, if you made it this far, you now have a functional air sensing machine.

Reviewer: ```suggestion
Congratulations.
If you made it this far, you now have a functional air sensing machine.
```
[nit] grammatically incorrect

Let's create a dashboard for its measurements next.

## Create a dashboard

The [Viam TypeScript SDK](https://ts.viam.dev/) allows you to build custom web interfaces to interact with your machines.
For this project, you'll use it to build a page that displays air quality sensor data for a given location.
You'll host the website on Viam Apps.

The full code is available for reference on [GitHub](https://github.com/viam-labs/air-quality-fleet/blob/main/main.ts).

{{< alert title="Tip" color="tip" >}}
If you'd like to graph your data using a Grafana dashboard, try our [Visualize Data with Grafana tutorial](/tutorials/services/visualize-data-grafana/) next.
{{< /alert >}}

### Set up your TypeScript project

Complete the following steps on your laptop or desktop.

> You don't need to install or edit anything else on your machine's single-board computer (aside from `viam-server` which you already did); you'll be developing your TypeScript app from your personal computer and deploying it to Viam.

Author: ```suggestion
You don't need to install or edit anything else on your machine's single-board computer (aside from `viam-server` which you already did); you'll be developing your TypeScript app from your personal computer and hosting it with Viam.
```

Paste the following code into <file>main.ts</file>:

```typescript {class="line-numbers linkable-line-numbers"}
// Air quality dashboard

import * as VIAM from "@viamrobotics/sdk";
import { BSON } from "bson";
import Cookies from "js-cookie";

let apiKeyId = "";
let apiKeySecret = "";
let hostname = "";
let machineId = "";

async function main() {
  const opts: VIAM.ViamClientOptions = {
    serviceHost: "https://app.viam.com",
    credentials: {
      type: "api-key",
      payload: apiKeySecret,
      authEntity: apiKeyId,
    },
  };

  // <Insert data client and query code here in later steps>

  // <Insert HTML block code here in later steps>
}

// <Insert getLastFewAv function definition here in later steps>

document.addEventListener("DOMContentLoaded", async () => {
  machineId = window.location.pathname.split("/")[2];
  ({
    id: apiKeyId,
    key: apiKeySecret,
    hostname: hostname,
  } = JSON.parse(Cookies.get(machineId)!));

  main().catch((error) => {
    console.error("encountered an error:", error);
  });
});
```

{{% /tablestep %}}
{{% tablestep number=2 %}}

For developing your app on localhost, **add the same information to your browsers local storage**.

> Navigate to [Camera Viewer](https://air-quality_naomi.viamapplications.com/) and log in, then select your development machine.

Reviewer: Same comment as other PR: It's confusing to tell them to seemingly go look at cameras without explanation of what this link is for. But sounds like it's about to change so can disregard so long as it'll get fixed.

  cookieSetterCode +=
    "// Copy and paste this entire script into your browser console when on localhost\n\n";

  // Process each cookie
  let cookieCount = 0;
  currentCookies.forEach((cookie) => {
    if (cookie.trim()) {
      // Extract name and value from the cookie
      const [name, value] = cookie.trim().split("=");

      // Add code to set this cookie
      cookieSetterCode += `document.cookie = "${name}=${value}; path=/";\n`;
      cookieCount++;
    }
  });

  // Add summary comment
  cookieSetterCode += `\nconsole.log("Set ${cookieCount} cookies on localhost");\n`;

  // Display the generated code
  console.log(cookieSetterCode);

  // Create a textarea element to make copying easier
  const textarea = document.createElement("textarea");
  textarea.value = cookieSetterCode;
  textarea.style.position = "fixed";
  textarea.style.top = "0";
  textarea.style.left = "0";
  textarea.style.width = "100%";
  textarea.style.height = "250px";
  textarea.style.zIndex = "9999";
  document.body.appendChild(textarea);
  textarea.focus();
  textarea.select();
}

// Execute the function
generateCookieSetterScript();
```

{{% /tablestep %}}
{{% tablestep number=3 %}}

> Copy the resulting script. It will look like this:

Reviewer: ```suggestion
Copy the resulting script.
It will look like this:
```
[nit]

```js {class="line-numbers linkable-line-numbers" data-line=""}
// Cookie setter script for localhost
// Copy and paste this entire script into your browser console when on localhost

document.cookie = "<SECRET COOKIE INFO>; path=/";
document.cookie = "machinesWhoseCredentialsAreStored=<MACHINE ID>; path=/";

console.log("Set 2 cookies on localhost");
```

{{% /tablestep %}}
{{% tablestep number=4 %}}

Run the following command to serve the app you are building:

```sh {class="command-line" data-prompt="$" data-output="2-10"}
npm run start
```

Open the app in your browser at `http://127.0.0.1:8000/`.

Then, open developer tools, go to the console and paste the copied JavaScript code to set your cookies.

{{% /tablestep %}}
{{< /table >}}

### Add functionality to your code

Now that you have the connection code, you are ready to add code that establishes a connection from the computer running the code to the Viam Cloud where the air quality sensor data is stored.

{{< table >}}
{{% tablestep number=1 %}}

> You'll first create a client to obtain the organization and location ID. Then you'll get a `dataClient` instance which accesses all the data in your location, and then query this data to get only the data tagged with the `air-quality` tag you applied with your data service configuration.

Reviewer: ```suggestion
You'll first create a client to obtain the organization and location ID.
Then you'll get a `dataClient` instance which accesses all the data in your location, and then query this data to get only the data tagged with the `air-quality` tag you applied with your data service configuration.
```
[nit] newline
Also is "get" the right word to use for what you do with dataclient? Should it be "establish" or "instantiate" or something more precise, or is that too pedantic?

// Get the average of the last five readings from a given sensor
async function getLastFewAv(all_measurements: any[], machineID: string) {
  // Get just the data from this machine
  let measurements = new Array();
  for (const entry of all_measurements) {
    if (entry.robot_id == machineID) {
      measurements.push({
        PM25: entry.data.readings["pm_2.5"],
        time: entry.time_received,
      });
    }
  }

  // Sort the air quality data from this machine
  // by timestamp
  measurements = measurements.sort(function (a, b) {
    let x = a.time.toString();
    let y = b.time.toString();
    if (x < y) {
      return -1;
    }
    if (x > y) {
      return 1;
    }
    return 0;
  });

  // Add up the last 5 readings collected.
  // If there are fewer than 5 readings, add all of them.
  let x = 5; // The number of readings to average over
  if (x > measurements.length) {
    x = measurements.length;
  }
  let total = 0;
  for (let i = 1; i <= x; i++) {
    const reading: number = measurements[measurements.length - i].PM25;
    total += reading;
  }
  // Return the average of the last few readings
  return total / x;
}
```

{{% /tablestep %}}
{{% tablestep number=3 %}}

> Now that you've defined the function to sort and average the data for each machine, you're done with all the `dataClient` code.

Reviewer: Should a bunch of the instances of "each machine" here and later, including in code comments, actually be "the machine" since the app only connects to one machine at a time?

Author: going to keep as is because the code still does that and if you read it as each machine you deploy individually i think it's ok. Hopefully soon there'll be acess to multiple machines anyway

## Requirements

You can create one or more machines to measure air quality.
For each machine, you need the following hardware:

- One [SDS011 Nova PM sensors](https://www.amazon.com/SDS011-Quality-Detection-Conditioning-Monitor/dp/B07FSDMRR5)
  - If you choose to use a different air quality sensor, you may need to [create your own module](/operate/get-started/other-hardware/) implementing the [sensor API](/operate/reference/components/sensor/#api) for your specific hardware.
- A single-board computer (SBC) [capable of running `viam-server`](https://docs.viam.com/installation/)
- An appropriate power supply

## Set up one device for development

In this section we'll set up one air sensing machine as our development device.

### Create your machine

{{< table >}}
{{% tablestep number=1 %}}

Navigate to [Viam](https://app.viam.com) in a web browser.
Create an account and log in.

{{% /tablestep %}}
{{% tablestep number=2 %}}

> Click the dropdown in the upper-right corner of the **FLEET** page and use the **+** button to create a new {{< glossary_tooltip term_id="organization" text="organization" >}} for your air quality machine company.

Author: ```suggestion
Click the dropdown in the upper-right corner of the **FLEET** page and use the **+** button to create a new {{< glossary_tooltip term_id="organization" text="organization" >}}.

 This tutorial walks you through how to set up your fleet as though you were creating a company selling air quality machines.
 If you already have a different suitable organization, you can use that instead.
```

let apiKeyId = "";
let apiKeySecret = "";
let host = "";
let machineId = "";

async function main() {
  const opts: VIAM.ViamClientOptions = {
    serviceHost: "https://app.viam.com",
    credentials: {
      type: "api-key",
      payload: apiKeySecret,
      authEntity: apiKeyId,
    },
  };

  // <Insert data client and query code here in later steps>

  // <Insert HTML block code here in later steps>
}

// <Insert getLastFewAv function definition here in later steps>

document.addEventListener("DOMContentLoaded", async () => {
  // Extract the machine identifier from the URL
  let machineCookieKey = window.location.pathname.split("/")[2];
  ({
    apiKey: { id: apiKeyId, key: apiKeySecret },
    machineId: machineId,
    hostname: host,
  } = JSON.parse(Cookies.get(machineCookieKey)!));
  main().catch((error) => {
    console.error("encountered an error:", error);
  });
  console.log(apiKeyId, apiKeySecret, host, machineId);
});
```

### Local development

For developing your application on localhost:

1. Run the following command to serve the application you are building:

   ```sh {class="command-line" data-prompt="$" data-output="2-10"}
   npm run start
   ```

>    Open the application in your browser at `http://localhost:8000/`.

Author: the npm run start command will still run on port 8000. the cli will then add more and proxy to 8012. 
However they actually don't need to open the application here. 
```suggestion
```

Navigate to one of the locations and create a machine.
Select the part status dropdown to the right of your machine's name on the top of the page.

Click the copy icon next to **Machine cloud credentials**.
Paste the machine cloud credentials into a file on your hard drive called FILE>viam.json</FILE>.

{{< alert title="Tip: Fleet management API" color="tip" >}}
You can create locations and machines programmatically, with the [Fleet management API](/dev/reference/apis/fleet/).
{{< /alert >}}

{{% /tablestep %}}
{{% tablestep number=4 %}}

**Run the preinstall script** without options and it will attempt to auto-detect a mounted root filesystem (or for Raspberry Pi, bootfs) and also automatically determine the architecture.

```sh {class="command-line" data-prompt="$"}
sudo ./preinstall.sh
```

Follow the instructions and provide the <FILE>viam-defaults.json</FILE> file and the machine cloud credentials file when prompted.

{{% /tablestep %}}
{{< /table >}}

That's it!
Your device is now provisioned and ready for your end user!

Having trouble?
See [Provisioning](/manage/fleet/provision/setup/) for more information and troubleshooting.

<div id="emailform"></div>

> 

Author: no - will leave this in for now

   ![Air quality dashboard in a web browser with PM2.5 readings from three different sensor machines displayed.](/tutorials/air-quality-fleet/three-sensor-dash.png)

Great work.
You've learned how to configure a machine and you can view its data in a custom TypeScript dashboard.

### Deploy the app to Viam apps

> Let's deploy this dashboard so you don't have to run it locally.

Reviewer: As said elsewhere, I think we use "deploy" to mean a lot of different things, so it'd be good to be more specific here. Without digging, not sure exactly what it should say but maybe:
```suggestion
Let's deploy this dashboard as a Viam-hosted app so you don't have to run it locally.
```

    }
  ]
}
```

In the [Viam app](https://app.viam.com), navigate to your organization settings through the menu in upper right corner of the page.
Find the **Public namespace** and copy that string.
Replace `<your-namespace>` with your public namespace.

{{% /tablestep %}}
{{% tablestep number=2 %}}

**Register your module** with Viam:

```sh {class="command-line" data-prompt="$" data-output="3-10"}
viam module create --name="air-quality" --public-namespace="your-namespace"
```

{{% /tablestep %}}
{{% tablestep number=3 %}}

**Package your static files and your <FILE>meta.json</FILE> file and upload them** to the Viam Registry:

```sh {class="command-line" data-prompt="$" data-output=""}
npm run build
tar -czvf module.tar.gz static meta.json
viam module upload --upload=module.tar.gz --platform=any --version=0.0.1
```

For subsequent updates run these commands again with an updated version number.

{{% /tablestep %}}
{{% tablestep number=4 %}}

**Try your app** by navigating to:

```txt
https://air-quality_your-public-namespace.viamapplications.com
```

Log in and select your development machine.
Your dashboard should now load your data.

{{% /tablestep %}}
{{< /table >}}

> ## Organizing devices for third-party usage

Reviewer: ```suggestion
## Organizing devices for third-party usage

The following example shows how you can use {{< glossary_tooltip term_id="organization" text="organizations" >}} and {{< glossary_tooltip term_id="location" text="locations" >}} to provide users access to the right groups of machines.
```
I recommend introducing this example in some way; this is just off the cuff

For more information, see [Fleet Management](/manage/reference/organize/) and [provisioning](/manage/fleet/provision/setup/).

### Organize your fleet

If you want to follow along, create the following locations:

- `Antonia's Home`
- `RobotsRUs`

For `RobotsRUs` crate two sublocations:

1. Add a new location called `Oregon Office` using the same **Add location** button.
1. Then, find the **New parent location** dropdown on the Oregon Office page.
1. Select **RobotsRUs** and click **Change**.

Repeat to add the New York office: Add a new location called `New York Office`, then change its parent location to **RobotsRUs**.

{{<imgproc class="imgzoom" src="/tutorials/air-quality-fleet/locations-done.png" resize="x900" declaredimensions=true alt="The New York Office fleet page. The left Locations navigation panel lists Antonia's Home and RobotsRUs, with New York Office and Oregon Office nested inside RobotsRUs." style="width:600px"  >}}

## Getting machines ready for third parties

> Continuing with our fictitious company, let's assume you want to ship air sensing machines to customers as ready-to-go as possible.

Reviewer: [opt] Awk, because "as" can be read different ways until you read the rest of the sentence. Consider restructuring.

In other words, you want to provision devices.

Before an air sensing machine leaves your factory, you'd complete the following steps:

> 1. You'd flash the single-board computer with an operating system
> 2. You'd install `viam-agent`

Reviewer: ```suggestion
1. You'd flash the single-board computer with an operating system.
2. You'd install `viam-agent`.
```
[nit] Consistency
Also consider linking to provisioning/viam-agent

> 3. You'd provide a machine configuration template, a fragment.

Reviewer: ```suggestion
3. You'd provide a machine configuration template: a _{{< glossary_tooltip term_id="fragment" text="fragment" >}}_.
```
This is the first mention of fragments on the page now.
The comma is grammatically wonky here.

Once a customer receives your machine, they will:

1. Plug it in and turn it on.

> 2. `viam-agent` will start a WiFi network

Reviewer: ```suggestion
2. `viam-agent` will start a WiFi network.
```

3. The customer uses another device to connect to the machine's WiFi network and the user gives the machine the password for their WiFi network.

> 4. The machine can now connect to the internet and complete setup based on the fragment it knows about.

Reviewer: "the fragment it knows about" is not clear and is a bit informal. Consider rewriting.

### Create the fragment for air sensing machines

> In this section you will create the {{< glossary_tooltip term_id="fragment" text="fragment" >}}, that is the configuration template that all other machines will use.

Reviewer: ```suggestion
In this section you will create the {{< glossary_tooltip term_id="fragment" text="fragment" >}}: the configuration template that all other machines will use.
```
As above, using a comma like this is incorrect. Could use a colon or an em dash.

1. Navigate to the **FLEET** page and go to the [**FRAGMENTS** tab](https://app.viam.com/fragments).

> 1. Click Create fragment.

Reviewer: ```suggestion
1. Click **Create fragment**.
```

1. Name the fragment `air-quality-configuration`.
1. Add the same components that you added to the development machine when you [set up one device for development](#set-up-one-device-for-development).

   As a shortcut, you can use the JSON mode on the machine you already configured and copy the machine's configuration to the fragment.

   {{< expand "Click here for info about the usb_interface value." >}}

If you only have one USB device plugged into each of your boards, the `usb_interface` value you configured in the sensor config is likely (conveniently) the same for all of your machines.

If not, you can use [fragment overwrite](/manage/fleet/reuse-configuration/#modify-fragment-settings-on-a-machine) to modify the value on any machine for which it is different.

{{< /expand >}}

> 1. Specify the version for the sds011 module.

Reviewer: ```suggestion
1. Specify the version for the `sds011` module.
```
[opt] I think?

   At the point of writing the version is `0.2.1`.
   Specifying a specific version or a specific minor or major version of a module will ensure that even if the module you use changes, your machines remain functional.

>    You can update your fragment at any point, any machines using it will update to use the new configuration.

Reviewer: ```suggestion
   You can update your fragment at any point, and any machines using it will update to use the new configuration.
```
or split the sentence

{{< alert title="Tip: Use the fragment on your development machine" color="tip" >}}

> To avoid differences between fragment and development machines, we recommend you remove the configured resources from the development machine and add the fragment you just created instead using the **+** button.

Reviewer: ```suggestion
To avoid differences between fragment and development machines, we recommend you remove the configured resources from the development machine, and instead use the **+** button to add the fragment you just created.
```
Slight legibility improvement. Could also try to restructure more significantly.

{{< /alert >}}

### Provision your machines

{{< table >}}
{{% tablestep number=1 %}}

For each machine, flash the operating system to the device's SD card.

> If you are using the Raspberry PI Imager, you **must customize at least the hostname** for the next steps to work.

Reviewer: ```suggestion
If you are using the Raspberry Pi Imager, you **must customize at least the hostname** for the next steps to work.
```
[nit]

Then run the following commands to download the preinstall script and make the script executable:

```sh {class="command-line" data-prompt="$"}
wget https://storage.googleapis.com/packages.viam.com/apps/viam-agent/preinstall.sh
chmod 755 preinstall.sh
```

{{% /tablestep %}}
{{% tablestep number=2 %}}

Create a file called <FILE>viam-defaults.json</FILE> with the following configuration:

```json {class="line-numbers linkable-line-numbers"}
{
  "network_configuration": {
    "manufacturer": "Pollution Monitoring Made Simple",
    "model": "v1",
    "fragment_id": "<FRAGMENT-ID>",
    "hotspot_prefix": "air-quality",
    "hotspot_password": "WeLoveCleanAir123"
  }
}
```

Replace `"<FRAGMENT-ID>"` with the fragment ID from your fragment.

{{% /tablestep %}}
{{% tablestep number=3 %}}

In [Organize your fleet](#organize-your-fleet) you created several locations.
Navigate to one of the locations and create a machine.
Select the part status dropdown to the right of your machine's name on the top of the page.

Click the copy icon next to **Machine cloud credentials**.

> Paste the machine cloud credentials into a file on your harddrive called FILE>viam.json</FILE>.

Reviewer: ```suggestion
Paste the machine cloud credentials into a file on your hard drive called FILE>viam.json</FILE>.
```

{{< alert title="Tip: Fleet management API" color="tip" >}}
You can create locations and machines programmatically, with the [Fleet management API](/dev/reference/apis/fleet/).
{{< /alert >}}

{{% /tablestep %}}
{{% tablestep number=4 %}}

**Run the preinstall script** without options and it will attempt to auto-detect a mounted root filesystem (or for Raspberry Pi, bootfs) and also automatically determine the architecture.

```sh {class="command-line" data-prompt="$"}
sudo ./preinstall.sh
```

Follow the instructions and provide the <FILE>viam-defaults.json</FILE> file and the machine cloud credentials file when prompted.

{{% /tablestep %}}
{{< /table >}}

> That's it! Your device is now provisioned and ready for your end user!

Reviewer: ```suggestion
That's it!
Your device is now provisioned and ready for your end user!
```
[nit]

> Having trouble? See [Provisioning](/manage/fleet/provision/setup/) for more information and troubleshooting.

Reviewer: ```suggestion
Having trouble?
See [Provisioning](/manage/fleet/provision/setup/) for more information and troubleshooting.
```
[nit]
